### PR TITLE
Addition of query based evaluator

### PR DESCRIPTION
**Pull request by jdraymon**
Resolves Issue #6593

### CODE REVIEW

**Comment by jdraymon** [comment]
NOTE! Tests are currently broken but I wanted to get some initial feedback on the approach in case I need to rework the alert system

**Comment by jdraymon** [comment]
Tests fixed, but no new ones added yet; still waiting on feedback of approach

**Comment by jdraymon** [comment]
@torkelo If you're good with this approach I'd like to do the remaining test work to get this PR ready for production

File: pkg/services/alerting/conditions/evaluator.go [hunk 1]
```diff
@@ -6,20 +6,27 @@ import (
...
 
 var (
 	defaultTypes []string = []string{"gt", "lt"}
+	queryTypes   []string = []string{"above_metric", "below_metric"}
 	rangedTypes  []string = []string{"within_range", "outside_range"}
 )
 
 type AlertEvaluator interface {
-	Eval(reducedValue null.Float) bool
+	Update(ctx *alerting.EvalContext) error
```
Comment (torkelo) [inline pkg/services/alerting/conditions/evaluator.go]: can you describe the need for adding this function to the interface?

Comment (jdraymon) [inline pkg/services/alerting/conditions/evaluator.go]: My initial implementation did without, but then I was re-issuing the comparison query for every series returned by the initial query.  Alternatively, the `Eval` interface could be rewritten such that it takes the full result set and a reducer, returning an EvalMatch struct. That seemed a more destructive change, so I avoided it at the time

**Comment by torkelo** [comment]
I have now looked at this in more detail. 

Not sure about this approach of complicating the query condition type with a subordinateQuery & reducer. I imagined this would use a new condition type or make a query condition be able to use the result of another condition. 

Sort of like this, but not sure about this either :(  tricky.
image

**Comment by jdraymon** [comment]
Well the subordinate query / reducer is a temporary conceit on the front end to make the current angular directives work properly; It's likely I could find a way to do without them and clean up that object. On the back end, those are handled purely as parameters to the evaluator, and I mostly avoided polluting the query space

I like the code model of your newer proposal, but to me the UX doesn't feel as natural as the original ticket... as you said, tricky :)

**Comment by torkelo** [comment]
Yea, agree, my proposal is not great, because in somecases you want a threshold or expression as well (a+b), greater than 10 for ecample

**Comment by torkelo** [comment]
Closing this as we need to redesign alerting conditions in some fashion to support this properly. And the approach taken in this PR does not really seem like a long or short term way forward.

**Comment by jdraymon** [comment]
@torkelo I have another method in progress that I have an intern working on, wherein I add a new query condition (MultiPartQueryCondition) - would that approach be acceptable? if you have a design in mind we'd be happy to implement it that way, but some direction would be helpful.  I can open the in progress version (Go done, Angular pending) if you'd like to review that.

**Comment by torkelo** [comment]
Early review is always great. A new condition could work but I worry that it will introduce to much code and UI duplication.

But beyond the approach I suggested above where a query condition can store a value and use a stored value (which feels unintuitive) I don’t have a good design in mind.

**Comment by maidirepelle** [comment]
> I have now looked at this in more detail.
> 
> Not sure about this approach of complicating the query condition type with a subordinateQuery & reducer. I imagined this would use a new condition type or make a query condition be able to use the result of another condition.
> 
> Sort of like this, but not sure about this either :( tricky.
> image

@torkelo how to configure a subordinateQuery with Grafana 6.5.0 ? Seems impossible to obtain the 'given' condition.

**Pull request closed without merging**